src/theory/builtin/abstract_type.h

```diff
@@ -0,0 +1,56 @@
+/******************************************************************************
+ * Top contributors (to current version):
+ *   Andrew Reynolds
+ *
+ * This file is part of the cvc5 project.
+ *
+ * Copyright (c) 2009-2022 by the authors listed in the file AUTHORS
+ * in the top-level source directory and their institutional affiliations.
+ * All rights reserved.  See the file COPYING in the top-level source
+ * directory for licensing information.
+ * ****************************************************************************
+ *
+ * A class for representing abstract types.
+ */
+
+#include "cvc5_public.h"
+
+#ifndef CVC5__THEORY__BUILTIN__ABSTRACT_TYPE_H
+#define CVC5__THEORY__BUILTIN__ABSTRACT_TYPE_H
+
+#include "expr/kind.h"
+
+namespace cvc5::internal {
+
+/**
+ */
+class AbstractType
+{
+ public:
+  AbstractType(Kind k);
+  AbstractType(const AbstractType& op);
+
+  /** return the kind of the current object */
+  Kind getKind() const;
+
+  bool operator==(const AbstractType& op) const;
+
+ private:
+  AbstractType();
+  /** a kind */
+  Kind d_kind;
+}; /* class AbstractType */
+
+std::ostream& operator<<(std::ostream& out, const AbstractType& op);
+
+/**
+ * Hash function for the AbstractType objects.
+ */
+struct AbstractTypeHashFunction
+{
+  size_t operator()(const AbstractType& op) const;
+}; /* struct AbstractTypeHashFunction */
```

Comment:
This should be defined the standard way, with `std::hash<AbstractType>` in namespace std.

Reply:
What do you suggest doing here? At the moment I am following the convention from other helper classes (e.g. `UninterpretedSortValueHashFunction`, `BitVectorExtractHashFunction`, etc.).

Reply:
I quickly tried and it seems that this doesn't play well at all with the kinds DSL, or rather, with the mkmetakind script. Leave it as is for now, we should convert the hash functions when we refactor that script.